internal/storage/bucket_handle.go

```diff
@@ -319,12 +319,29 @@ func (b *bucketHandle) UpdateObject(ctx context.Context, req *gcs.UpdateObjectRe
 func (b *bucketHandle) ComposeObjects(ctx context.Context, req *gcs.ComposeObjectsRequest) (o *gcs.Object, err error) {
 	dstObj := b.bucket.Object(req.DstName)
 
-	if req.DstGenerationPrecondition != nil && req.DstMetaGenerationPrecondition != nil {
-		dstObj = dstObj.If(storage.Conditions{GenerationMatch: *req.DstGenerationPrecondition, MetagenerationMatch: *req.DstMetaGenerationPrecondition})
-	} else if req.DstGenerationPrecondition != nil {
-		dstObj = dstObj.If(storage.Conditions{GenerationMatch: *req.DstGenerationPrecondition})
-	} else if req.DstMetaGenerationPrecondition != nil {
-		dstObj = dstObj.If(storage.Conditions{MetagenerationMatch: *req.DstMetaGenerationPrecondition})
+	dstObjConds := storage.Conditions{}
+	if req.DstMetaGenerationPrecondition != nil {
+		dstObjConds.MetagenerationMatch = *req.DstMetaGenerationPrecondition
+	}
+	// DstGenerationPrecondition or DoesNotExist should be set in dstObj
+	// preconditions to make requests Idempotent.
+	// https://github.com/GoogleCloudPlatform/gcsfuse/blob/7ad451c6f2ead7992e030503e5b66c555b2ebf71/vendor/cloud.google.com/go/storage/copy.go#L230
+	// When DstGenerationPrecondition is 0, set DoesNotExist as true.
+	// https://github.com/GoogleCloudPlatform/gcsfuse/blob/7ad451c6f2ead7992e030503e5b66c555b2ebf71/vendor/cloud.google.com/go/storage/storage.go#L1710
```

Comment:
we shouldn't add too many comments in the code. If you click on dstObjConds.DoesNotExist its clear when to set

Reply:
removed this comment.